### PR TITLE
Autotune API

### PR DESCRIPTION
**Pull request by mjayasur**

### CODE REVIEW

File: scvi/autotune/_autotune.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            _model.save(os.path.join(checkpoint_dir, "checkpoint"))
+
+    def run(
+        self,
+        metric,
+        scheduler,
+        mode="min",
+        name="scvi-experiment",
+        num_samples=10,
+    ):
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:96]: **kwargs to tune.run

File: scvi/autotune/_autotune.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    def __init__(
+        self,
+        adata,
+        model,
+        training_metrics: list = ["elbo_validation"],
+        metric_functions: dict = {},
+        model_hyperparams: dict = {},
+        trainer_hyperparams: dict = {},
+        plan_hyperparams: dict = {},
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:38]: Remove mutable defaults, should be `Optional[dict] = None`

File: scvi/autotune/_autotune.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        self.reporter = CLIReporter(
+            metric_columns=training_metrics + list(self.metric_functions.keys())
+        )
+        self.config = {}
+        for d in [model_hyperparams, trainer_hyperparams, plan_hyperparams]:
+            if d is not None:
+                self.config.update(d)
+        self.num_epochs = num_epochs
+
+    def _scvi_trainable(self, config, checkpoint_dir=None):
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:58]: I'd call this just `_trainable` or something

File: scvi/autotune/_autotune.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+            max_epochs=self.num_epochs,
+        )
+
+    def run(
+        self,
+        metric,
+        scheduler,
+        mode="min",
+        name="scvi-experiment",
+        num_samples=10,
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:93]: typing

File: scvi/autotune/_autotune.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
+from ray import tune
+from ray.tune import CLIReporter
+
+from ._callbacks import ModelSave, _TuneReportMetricFunctionsCallback
+
+
+class Autotune:
+    """
+
+    Hyperparameter tuning for SCVI using Ray Tune.
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:10]:
```suggestion
    Hyperparameter tuning using Ray Tune.
```

File: scvi/autotune/_autotune.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            Metric to optimize over in self.metrics or from self.training_funcs
+        scheduler
+            Ray tune scheduler for trials (e.g. ASHA).
+        mode
+            "min" or "max" to maximize or minimize the objective metric
+        name
+            Name of this experiment.
+        num_samples
+            Number of times to sample hyperparameters from the configuration space.
+
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:111]: document kwargs

File: scvi/autotune/_autotune.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        self,
+        metric,
+        scheduler,
+        mode="min",
+        name="scvi-experiment",
+        num_samples=10,
+        **kwargs,
+    ):
+        """
+        Run hyper parameter tuning experiment.
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:97]: some description that this is a wrapper of tune.run

File: scvi/autotune/_autotune.py (hunk 6 above)
Comment (adamgayoso) [reply scvi/autotune/_autotune.py:111]: also a returns part describing the outputs

File: scvi/autotune/_autotune.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                ),
+            ],
+            check_val_every_n_epoch=1,
+            max_epochs=self.num_epochs,
+        )
+
+    def run(
+        self,
+        metric,
+        scheduler,
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:90]: should there be a default scheduler? So default is None and if None, use Asha...

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/autotune/_autotune.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+            check_val_every_n_epoch=1,
+            max_epochs=self.num_epochs,
+        )
+
+    def run(
+        self,
+        metric: str = None,
+        scheduler: TrialScheduler = None,
+        mode: str = "min",
+        name: str = "scvi-experiment",
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:97]: let's add a default search alg (hyperopt) default None here

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/autotune/_wrappers.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    metrics = [
+        "elbo_validation",
+        "reconstruction_loss_validation",
+    ]
+    metric_functions = {}
+    model_config = {
+        "dropout_rate": loguniform(1e-4, 1e-1),
+        "n_layers": choice([i for i in range(1, 5)]),
+        "n_hidden": choice([64, 128, 256]),
+        "n_latent": choice([i for i in range(5, 15)]),
```
Comment (adamgayoso) [inline scvi/autotune/_wrappers.py:33]: `batch_size: choice([128, 256, 512, 1024])`

File: scvi/autotune/_autotune.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        if not scheduler:
+            scheduler = ASHAScheduler(max_t=2, grace_period=1, reduction_factor=2)
+
+        if not search_alg:
+            search_alg = HyperOptSearch()
+
+        if not resources_per_trial:
+            if torch.cuda.is_available():
+                resources_per_trial = {"cpu": 1, "gpu": 1}
```
Comment (adamgayoso) [inline scvi/autotune/_autotune.py:150]: let's see if we can run multiple trials on the same gpu

**Comment by mjayasur** [comment]
mmd default and FIFO

**Comment by mjayasur** [comment]
stochastic_weight_avg

File: scvi/autotune/_callbacks.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    def on_validation_epoch_end(self, trainer, pl_module, outputs=None):
+        if trainer.running_sanity_check:
+            return
+        step = f"epoch={trainer.current_epoch}-step={trainer.global_step}"
+        with tune.checkpoint_dir(step=step) as checkpoint_dir:
+            self.model.save(checkpoint_dir + "/checkpoint")
+
+
+class _TuneReportMetricFunctionsCallback(TuneCallback):
+    def __init__(
```
Comment (adamgayoso) [inline scvi/autotune/_callbacks.py:20]: add documentation to this callback

File: scvi/autotune/_callbacks.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        self,
+        metrics=None,
+        metric_functions=None,
```
Comment (adamgayoso) [inline scvi/autotune/_callbacks.py:23]: add somewhere how metric functions need to be formatted

File: scvi/autotune/_metrics.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        mmd_loss += _compute_fast_mmd(z0, z1)
+
+    return mmd_loss
+
+
+def mmd(model):
+    return _compute_mmd_loss(
+        torch.tensor(model.adata.obsm["X_scVI"]),
+        torch.tensor(model.adata.obs._scvi_batch),
+    )
```
Comment (adamgayoso) [inline scvi/autotune/_metrics.py:106]: remove for now

File: scvi/autotune/_callbacks.py (hunk 13 above)
Comment (adamgayoso) [reply scvi/autotune/_callbacks.py:23]: also need to check that their signature is correct

**Comment by jjhong922** [comment]
Closed in favor of https://github.com/scverse/scvi-tools/pull/1565
